nucypher/crypto/kits.py

```diff
@@ -15,76 +15,112 @@
+            raise ValueError("Cannot include a verifying key. None was included in the MessageKit.")
+
+        if include_sender_verifying_key:
+            header = KIT_INCLUDES_SIGNATURE if include_signature else KIT_SIGNED_BUT_NO_SIGNATURE
+        else:
+            header = KIT_NOT_SIGNED
 
-        # Then, before the ciphertext, we see if we're including alice's public key.
+        # We include the header first, next the capsule.
+        as_bytes = header + bytes(self.capsule)
+
+        # Then, before the ciphertext, we see if we're including the sender's public key.
         # We want to put that first because it's typically of known length.
-        if include_alice_pubkey and self.sender_verifying_key:
+        if include_sender_verifying_key:
             as_bytes += bytes(self.sender_verifying_key)
 
+        if include_signature:
+            as_bytes += bytes(self.signature)
+
         as_bytes += self.ciphertext
-        return as_bytes
 
-    @property
-    def signature(self):
-        return self._signature
+        return as_bytes
 
     def __bytes__(self):
-        return bytes(self.capsule) + self.ciphertext
+        it_has_a_key = self.sender_verifying_key is not None
+        return self.to_bytes(include_sender_verifying_key=it_has_a_key,
+                             include_signature=False)
+
+    def to_base64(self):
+        return base64.b64encode(bytes(self)).decode("utf-8")
 
+    @classmethod
+    def from_bytes(cls, some_bytes: bytes):
+        header, capsule, remainder = cls.base_splitter(some_bytes, return_remainder=True)
+        if header == KIT_NOT_SIGNED:
+            sender_verifying_key = None
+            signature = NOT_SIGNED
+            ciphertext = remainder
+        elif header == KIT_SIGNED_BUT_NO_SIGNATURE:
+            sender_verifying_key, ciphertext = key_splitter(remainder, return_remainder=True)
+            signature = NOT_SIGNED
+        elif header == KIT_INCLUDES_SIGNATURE:
+            splitter = key_splitter + signature_splitter
+            sender_verifying_key, signature, ciphertext = splitter(remainder, return_remainder=True)
+
+        message_kit = cls(capsule=capsule,
+                          sender_verifying_key=sender_verifying_key,
+                          ciphertext=ciphertext,
+                          signature=signature)
+        return message_kit
 
-class UmbralMessageKit(MessageKit):
+    @classmethod
+    def from_base64(cls, some_b64_str: str):
+        some_bytes = base64.b64decode(some_b64_str)
+        message_kit = cls.from_bytes(some_bytes)
+        return message_kit
 
-    return_remainder_when_splitting = True
-    splitter = capsule_splitter + key_splitter
+    def __eq__(self, other):
+        eq_conditions = (self.ciphertext == other.ciphertext,
+                         self.capsule == other.capsule,
+                         self.sender_verifying_key == other.sender_verifying_key)
```

Comment:
Do we need to match on `signature` as well or is it purposely ignored?

Reply:
It was on purpose, and loosely related to this interchange between @jMyles and me:

> dnunez: something that I'd like your opinion here is the relation between the MessageKit and it's signature. It's very weird since (1) the signature is included in the constructor, but (2) it's not serialized together with the MessageKit (nor deserialized obviously), and (3) it's passed separately in the output of encrypt_and_sign() (and therefore in the output of Enrico.encrypt())
> jmyles: I agree that this area needs some love (in fact see my PR, which makes potentially useful changes - I'm still hoping to get that merged along with the demo material).  I don't per se agree that including the signature in the (or at least in some) constructor but then not serializing it as a separate component is strange.